packages/store/addon/-private/system/record-array-manager.js

```diff
@@ -8,347 +8,804 @@ import { get, set } from '@ember/object';
+    }
+  };
+
+  const destroy = function(entry) {
+    entry.destroy();
+  };
+
+  const remove = function(array, item) {
+    let index = array.indexOf(item);
+
+    if (index !== -1) {
+      array.splice(index, 1);
+      return true;
+    }
+
+    return false;
+  };
+
+  const updateInternalModelsForLiveRecordArray = function(array, internalModels) {
     let modelsToAdd = [];
+    let modelsToRemove = [];
+
     for (let i = 0; i < internalModels.length; i++) {
       let internalModel = internalModels[i];
+      let isDeleted = internalModel.isHiddenFromRecordArrays();
       let recordArrays = internalModel._recordArrays;
-      if (recordArrays.has(array) === false) {
-        recordArrays.add(array);
-        modelsToAdd.push(internalModel);
+
+      if (!isDeleted && !internalModel.isEmpty()) {
+        if (!recordArrays.has(array)) {
+          modelsToAdd.push(internalModel);
+          recordArrays.add(array);
+        }
+      }
+
+      if (isDeleted) {
+        modelsToRemove.push(internalModel);
+        recordArrays.delete(array);
       }
     }
 
-    if (modelsToAdd.length) {
+    if (modelsToAdd.length > 0) {
       array._pushInternalModels(modelsToAdd);
     }
-  }
+    if (modelsToRemove.length > 0) {
+      array._removeInternalModels(modelsToRemove);
+    }
+  };
 
-  _didUpdateAll(modelName) {
-    let recordArray = this._liveRecordArrays[modelName];
-    if (recordArray) {
-      set(recordArray, 'isUpdating', false);
+  const removeInternalModelsFromAdapterPopulatedRecordArrays = function(internalModels) {
+    for (let i = 0; i < internalModels.length; i++) {
+      removeInternalModelFromAll(internalModels[i]);
     }
-  }
+  };
 
-  /**
-    Get the `RecordArray` for a modelName, which contains all loaded records of
-    given modelName.
-
-    @method liveRecordArrayFor
-    @param {String} modelName
-    @return {RecordArray}
-  */
-  liveRecordArrayFor(modelName) {
-    assert(
-      `recordArrayManger.liveRecordArrayFor expects modelName not modelClass as the param`,
-      typeof modelName === 'string'
-    );
-
-    let array = this._liveRecordArrays[modelName];
-
-    if (array) {
-      // if the array already exists, synchronize
-      this._syncLiveRecordArray(array, modelName);
-    } else {
-      // if the array is being newly created merely create it with its initial
-      // content already set. This prevents unneeded change events.
-      let internalModels = this._visibleInternalModelsByType(modelName);
-      array = this.createRecordArray(modelName, internalModels);
-      this._liveRecordArrays[modelName] = array;
+  const removeInternalModelFromAll = function(internalModel) {
+    const recordArrays = internalModel._recordArrays;
+
+    recordArrays.forEach(function(recordArray) {
+      recordArray._removeInternalModels([internalModel]);
+    });
+
+    recordArrays.clear();
+  };
+
+  associateWithRecordArray = function assocWithRecordArray(internalModels, array) {
```

Comment:
should we maybe name `assocWithRecordArray` something more like `legacyAssocWithRecordArray`